File: api/src/main/java/dev/adamsalves/ordertracker/order/OrderStatus.java
package dev.adamsalves.ordertracker.order;

import java.util.Set;

public enum OrderStatus {
    RECEBIDO,
    EM_PREPARO,
    SAIU_PARA_ENTREGA,
    ENTREGUE,
    CANCELADO;

    /**
     * The table is read by a switch rather than held in a field: a constant cannot reference its
     * siblings while they are still being constructed, and the method body runs long after that.
     *
     * <p>No status lists itself, so standing still is refused by the same table that refuses a jump
     * over a step. Adding a self-entry here would quietly allow it.
     */
    public Set<OrderStatus> allowedTransitions() {
        return switch (this) {
            case RECEBIDO -> Set.of(EM_PREPARO, CANCELADO);
            case EM_PREPARO -> Set.of(SAIU_PARA_ENTREGA, CANCELADO);
            case SAIU_PARA_ENTREGA -> Set.of(ENTREGUE, CANCELADO);
            case ENTREGUE, CANCELADO -> Set.of();
        };
    }

    public boolean canTransitionTo(OrderStatus target) {
        return allowedTransitions().contains(target);
    }
}

File: web/src/lib/orderStatus.ts
import type { OrderStatus } from "@/api/types";

/**
 * The one-way path. CANCELADO is missing on purpose: it leaves the path, it is not a step on it.
 */
export const FORWARD_STATUSES: readonly OrderStatus[] = [
  "RECEBIDO",
  "EM_PREPARO",
  "SAIU_PARA_ENTREGA",
  "ENTREGUE",
];

/** The enum values are Portuguese by requirement; these are the same values made readable. */
export const STATUS_LABELS: Record<OrderStatus, string> = {
  RECEBIDO: "Recebido",
  EM_PREPARO: "Em preparo",
  SAIU_PARA_ENTREGA: "Saiu para entrega",
  ENTREGUE: "Entregue",
  CANCELADO: "Cancelado",
};

/** What the action does, rather than what it lands on, since it labels a button. */
export const TRANSITION_LABELS: Record<OrderStatus, string> = {
  RECEBIDO: "Receber",
  EM_PREPARO: "Iniciar preparo",
  SAIU_PARA_ENTREGA: "Despachar para entrega",
  ENTREGUE: "Confirmar entrega",
  CANCELADO: "Cancelar pedido",
};

/**
 * Mirrors OrderStatus.allowedTransitions() on the API. Kept as a convenience so the screen only
 * offers what will be accepted — the server stays the authority and answers 409 either way.
 *
 * No status lists itself, so standing still is refused by the same table that refuses a jump.
 */
const TRANSITIONS: Record<OrderStatus, readonly OrderStatus[]> = {
  RECEBIDO: ["EM_PREPARO", "CANCELADO"],
  EM_PREPARO: ["SAIU_PARA_ENTREGA", "CANCELADO"],
  SAIU_PARA_ENTREGA: ["ENTREGUE", "CANCELADO"],
  ENTREGUE: [],
  CANCELADO: [],
};

export function allowedTransitions(
  status: OrderStatus,
): readonly OrderStatus[] {
  return TRANSITIONS[status];
}

export function isTerminal(status: OrderStatus): boolean {
  return TRANSITIONS[status].length === 0;
}
